### ISSUE TITLE
Django URLs with multiple placeholders cut off after first placeholder

### ISSUE THREAD

**Issue by Abdkhan14** [getsentry/sentry-python#2392]
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

1. The transaction is associated with loading a trace view. 
2. Here's the parent pageload: [link](https://sentry.sentry.io/performance/summary/?project=11276&query=&referrer=performance-transaction-summary&statsPeriod=24h&transaction=%2Fperformance%2Ftrace%2F%3AtraceSlug%2F&unselectedSeries=p100%28%29&unselectedSeries=avg%28%29)
3. Correct transaction summary: [link](https://sentry.sentry.io/performance/summary/?project=1&query=http.method%3AGET&referrer=performance-transaction-summary&statsPeriod=90d&transaction=%2Fapi%2F0%2Forganizations%2F%7Borganization_slug%7D%2Fevents-trace%2F%7Btrace_id%7D%2F&unselectedSeries=p100%28%29&unselectedSeries=avg%28%29)
4. Renamed transaction summary: [link](https://sentry.sentry.io/performance/summary/?project=1&query=&referrer=performance-transaction-summary&statsPeriod=90d&transaction=%2Fapi%2F0%2Forganizations%2F%7Borganization_slug%7D%2F&unselectedSeries=p100%28%29&unselectedSeries=avg%28%29)
5. The change that introduced the bug: [link](https://github.com/getsentry/sentry-python/pull/1773/files)

### Expected Result

Transaction name is` /api/0/organizations/{organization_slug}/events-trace/{trace_id}/`

### Actual Result

Transaction name is renamed from `/api/0/organizations/{organization_slug}/events-trace/{trace_id}/` to `/api/0/organizations/{organization_slug}/`. 

### Product Area

Other

**Comment by smeubank** [comment]
https://github.com/getsentry/sentry-python/discussions/2416

also noted here in this discussion

**Comment by sentrivana** [comment]
Relevant/might be fixed by: https://github.com/getsentry/sentry-python/pull/2325

**Comment by sentrivana** [comment]
TL;DR: My proposed course of action for now:
* revert the change that introduced the regression, adding test cases to make sure it doesn't happen again https://github.com/getsentry/sentry-python/pull/2432
* reopen the original Django CMS issue https://github.com/getsentry/sentry-python/issues/1527 and revisit it at a later point
* consider whether the resolver makes sense as is because of the limitations outlined below

---

Some background:

The fundamental problem here is that we're trying to parse a non-regular language with regular expressions in the [resolver](https://github.com/getsentry/sentry-python/blob/b873a31fb432a8b0cb5adb74a64978a87a33f6d3/sentry_sdk/integrations/django/transactions.py#L39-L44). Any regex we come up with will never fully "work". There will always be counter examples of perfectly fine URL patterns that will not be parsed correctly.

To illustrate, consider these proposed candidates and URL patterns they won't work for (paste e.g. to https://regex101.com/ to play with these):

- `r"\(\?P<(\w+)>.*\)"`: This is the currently used regex that matches too much, so e.g. for a URL pattern such as `(?P<project>\w+)/product/(?P<pid>\w+)` it greedily matches the whole string instead of two individual groups. This is what leads to us essentially erasing stuff after the first matched group if there's any closing parenthesis anywhere further in the string.
- `r"\(\?P<(\w+)>[^\)]+\)+"`: This is the original regex, which, while working fine for the multiple named groups case, as soon as there's an extra closing parenthesis in an unfortunate place, will match too little: `(?P<project>\w+[()]+)/product/(?P<pid>\w+)` (this is a very simplified version of the URL pattern from https://github.com/getsentry/sentry-python/issues/1527).
- `r"\(\?P<(\w+)>.*?\)(?=/|\$|$)"`: This is a proposal that covers both the Django CMS case and the multiple named groups case, but comes with a baked in assumption that any named group will be neatly in its own "container" ending with a slash or end of string. If someone has some static stuff appended to a named group, e.g. `(?P<project>\w+)/product/(?P<pid>\d+)p`, the named group will not be captured.

While this list is definitely not exhaustive, the point I want to make is that _any_ regex we choose will have to come with some assumptions about what constitutes a "correct URL pattern", simply because regexes inherently can't solve this problem. In other words, any change we will make to the regex will introduce a regression.

**Comment by salomvary** [comment]
👍 for the proposed course of action.

Would supporting "nice" transaction names for `path`s only, while `re_path`s would be exposed raw?

**Comment by sentrivana** [comment]
@salomvary The problem is that `path`s can also contain "hidden" regexes, see e.g. https://github.com/getsentry/sentry-python/issues/2446.
